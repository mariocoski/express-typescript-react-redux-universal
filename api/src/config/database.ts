const env = require('../utils').env; 

module.exports = {
  development: {
    username: 'root',
    password: 'root',
    database: 'database_dev',
    host: '127.0.0.1',
    dialect: 'mysql'
  },
  test: {
    username: env('TEST_DB_USERNAME'),
    password: env('TEST_DB_PASSWORD'),
    database: env('TEST_DB_NAME'),
    dialect: 'sqlite'
  },
  production: {
    username: process.env.PROD_DB_USERNAME,
    password: process.env.PROD_DB_PASSWORD,
    database: process.env.PROD_DB_NAME,
    host: process.env.PROD_DB_HOSTNAME,
    dialect: 'mysql'
  }
}