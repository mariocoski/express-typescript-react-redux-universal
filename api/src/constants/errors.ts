export const EMAIL_IS_REQUIRED = 'Email is required';
export const EMAIL_IS_INVALID = "Invalid email";
export const EMAIL_ALREADY_IN_USE = 'That email address is already in use';
export const PASSWORD_IS_REQUIRED = 'Password is required';
export const PASSWORD_IS_TOO_SHORT = 'Password must be at least 6 characters long';
export const PASSWORD_CONFIRMATION_MUST_MATCH_PASSWORD = 'password_confirmation field must have the same value as the password field';
export const INVALID_CREDENTIALS = 'Wrong email and/or password';
export const USER_NOT_FOUND = 'We could not find your email address';
export const TOKEN_IS_REQUIRED = 'Token is required';
export const INVALID_PASSWORD_RESET_TOKEN = 'Invalid password reset token';
export const EXPIRED_PASSWORD_RESET_TOKEN = 'Token expired';
export const UNAUTHORIZED = 'Unauthorized';