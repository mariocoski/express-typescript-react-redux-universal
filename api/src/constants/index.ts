export const ONE_HOUR = 3600000;
export const TWO_HOURS = 7200000;