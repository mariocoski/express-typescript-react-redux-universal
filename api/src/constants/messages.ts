export const PASSWORD_CHANGE_SUCCESS = 'Password changed successfully. Please login with your new password';