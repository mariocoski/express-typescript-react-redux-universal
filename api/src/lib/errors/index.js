import BaseError from './BaseError';

import BadRequestError from './BadRequestError';

import UnauthorizedError from './UnauthorizedError';

import ForbiddenError from './ForbiddenError';

import NotFoundError from './NotFoundError';

export {
  BaseError,
  BadRequestError,
  UnauthorizedError,
  ForbiddenError,
  NotFoundError,
};
