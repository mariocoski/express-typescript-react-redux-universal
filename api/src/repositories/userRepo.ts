const db = require('../models');

export async function findUserByEmail(email: string) {
  return db.User.findOne({ where: { email } });
}

export async function findUserByToken(token: string){
  return db.User.findOne({ where: { password_reset_token: token } });
}

export async function createUser(values: Object, settings: Object = {}) {
  const options = { fields: 
    ['email','password', 'first_name', 'last_name', 'bio'],
    ...settings 
  };
  return db.User.create(values, options);
}

export default {
  findUserByToken,
  findUserByEmail,
  createUser
}
