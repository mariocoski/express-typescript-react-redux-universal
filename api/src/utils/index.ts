import {ROLE_USER, ROLE_PROFESSIONAL, ROLE_ADMIN, ROLE_SUPERADMIN} from '../constants/roles';
import {Request, Response} from 'express';
import {User} from '../models/user';

export function resolvePort(val: any) {
    const port = parseInt(val, 10);
  
    if (isNaN(port)) {
      return val;
    }
  
    if (port >= 0) {
      return port;
    }
    
    return 3000;
}



export const setUserInfo = (request) => {
  const getUserInfo = {
    _id: request._id,
    firstName: request.profile.firstName,
    lastName: request.profile.lastName,
    email: request.email,
    role: request.role
  };

  return getUserInfo;
};

export const getRole = (checkRole) => {
  let role;

  switch (checkRole) {
    case ROLE_SUPERADMIN: role = 4; break;
    case ROLE_ADMIN: role = 3; break;
    case ROLE_PROFESSIONAL: role = 2; break;
    case ROLE_USER: role = 1; break;
    default: role = 1;
  }

  return role;
};

export function isAuthorized(role: String){
  return function(req: Request,res: Response,next:Function){
    const user = req.user;
   
    User.findById(user._id, (err, foundUser:any)=>{
  
      if(err) {
        res.status(422).json({error: 'No user found'});
        return next(err);
      }
      if(getRole(foundUser.role) >= getRole(role)){
        return next();
      }
      res.status(401).json({error: "You are not authorized to see the content"});
      return next(err);
    });
   
    
  }
}


export function env(key): string {
  const myVal: string | undefined = process.env[key];
  if(myVal === undefined){
      throw new Error(`No ENV ${key} not found`);
  }
  return myVal;
}