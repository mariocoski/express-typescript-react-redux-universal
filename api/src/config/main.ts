import { env } from '../utils';
export default {
  mailgun_domain: env('MAILGUN_DOMAIN'),
  mailgun_api_key: env('MAILGUN_API_KEY')
};