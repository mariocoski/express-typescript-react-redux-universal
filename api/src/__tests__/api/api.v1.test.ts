require('dotenv').config();
import { generateToken,comparePassword} from '../../utils';
import {findUserById, findUserByEmail} from '../../repositories/userRepo';
import {UNAUTHORIZED,USER_DOES_NOT_EXIST,PASSWORD_IS_TOO_SHORT,EMAIL_IS_INVALID} from '../../constants/errors';
import config from '../../config/main';
import {expectError} from '../helpers';

const db = require('../../models');

describe('API V1', () => {

  const request = require('supertest');
  let app: any;

  beforeEach(async() => {
    await db.sequelize.sync({force:true});
    app = require('../../server');
  });

  afterEach(async() => {
    app.close();
  });

  it('should respond with 200 for health check', async () => {
    expect.assertions(1);
    const response = await request(app).get('/api/v1');
    expect(response.statusCode).toBe(200);
  });

  it('should respond with 404', async () => {
    expect.assertions(2);
    const response = await request(app).get('/not-existing-route');
    expect(response.statusCode).toBe(404);
    expect(response.text).toMatchSnapshot();
  });

  it('should respond with 401 when token is not present', async () => {
    const response = await request(app).get('/api/v1/profile');
    expect(response.statusCode).toBe(401);
    expect(response.text).toMatchSnapshot();
  });

  it('should respond with 401 when token is invalid', async() => {
    const response = await request(app)
                                  .get('/api/v1/profile')
                                  .set('Authorization', 'Bearer invalid-token');
    expect(response.statusCode).toBe(401);
    expect(response.text).toMatchSnapshot();
  });

  it('should respond with 400 when _id is not present in user data', async() => {
    const userIdWhichDoesNotExist = 999;
    const token = await generateToken({});
    const response = await request(app)
                                  .get('/api/v1/profile')
                                  .set('Authorization',`Bearer ${token}`);
    expect(response.statusCode).toBe(400);
  });

  it('should respond with 401 when token is valid, but user does not exist anymore', async() => {
    const userIdWhichDoesNotExist = 999;
    const token = await generateToken({_id: userIdWhichDoesNotExist});
    const response = await request(app)
                                  .get('/api/v1/profile')
                                  .set('Authorization',`Bearer ${token}`);
    expect(response.statusCode).toBe(401);
    expect(response.text).toMatchSnapshot();
  });

  it('should respond with 200 when token is valid and user exists', async() => {
    const validUser = {email:config.mailgun_test_recipient, password: 'password'};
    const foundUser = await db.User.create(validUser);
    const token = await generateToken({_id: foundUser.id});
    const response = await request(app)
                                  .get('/api/v1/profile')
                                  .set('Authorization',`Bearer ${token}`);
    expect(response.statusCode).toBe(200);
    expect(response.text).toMatch(config.mailgun_test_recipient);
  });

  it('should fail to update a profile if password is present but too short', async () => {
    const validUser = {email:config.mailgun_test_recipient, password: 'password'};
    const createdUser = await db.User.create(validUser);
    const token = await generateToken({_id: createdUser.id});

    const response = await request(app)
                            .patch(`/api/v1/profile/${createdUser.id}`)
                            .set('Authorization',`Bearer ${token}`)
                            .type('form')
                            .send({ password: 'short' });

    expectError(response,PASSWORD_IS_TOO_SHORT);
  });

  it('should fail to update user when email field is present but invalid', async () => {
    const validUser = {email:config.mailgun_test_recipient, password: 'password'};
    const createdUser = await db.User.create(validUser);
    const token = await generateToken({_id: createdUser.id});
    const response = await request(app) .patch(`/api/v1/profile/${createdUser.id}`)
                                        .set('Authorization',`Bearer ${token}`)
                                        .type('form')
                                        .send({ email: 'invalid@email'});
    expectError(response,EMAIL_IS_INVALID);
  });

  it('should update user when data are valid', async() => {

    const validUser = {email:config.mailgun_test_recipient, password: 'password'};
    const createdUser = await db.User.create(validUser);
    const token = await generateToken({_id: createdUser.id});
    const newData = { 
      email: 'newemail@email.com',
      password: 'password2',
      bio: 'Natural Born Killer',
      first_name: 'James',
      last_name: 'Bond'
    };
  
    const response = await request(app)
                                .patch(`/api/v1/profile/${createdUser.id}`)
                                .set('Authorization',`Bearer ${token}`)
                                .type('form')
                                .send(newData);
    const updatedUser = await findUserById(createdUser.id);
    const passwordMatch = await comparePassword(newData.password, updatedUser.password);
    expect(response.statusCode).toBe(200);
    expect(updatedUser.email).toBe(newData.email);
    expect(updatedUser.bio).toBe(newData.bio);
    expect(updatedUser.first_name).toBe(newData.first_name);
    expect(updatedUser.last_name).toBe(newData.last_name);
    expect(passwordMatch).toBe(true);
  });

});
