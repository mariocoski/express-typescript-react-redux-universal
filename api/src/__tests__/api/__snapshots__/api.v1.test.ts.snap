// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API V1 should respond with 401 when token is invalid 1`] = `"{\\"error\\":\\"Unauthorized\\"}"`;

exports[`API V1 should respond with 401 when token is not present 1`] = `"{\\"error\\":\\"Unauthorized\\"}"`;

exports[`API V1 should respond with 401 when token is valid, but user does not exist anymore 1`] = `"{\\"error\\":\\"Unauthorized\\"}"`;

exports[`API V1 should respond with 404 1`] = `"{\\"error\\":\\"Not found\\"}"`;
